### PR TITLE
check raftcmd term.

### LINKED ISSUE #1317: Make sure different stages of a kv command execute raft command in a same term.
Currently, a mutable kv command's execution is split into 3 stages: (Refer https://github.com/pingcap/tikv/issues/807)
- Stage1: Create snapshot (through Raft).
- Stage2: Execute txn command (on snapshot).
- Stage3: Write modifications (through Raft).

If leader transfers away then quickly transfers back between Stage1 and Stage3, another transaction may abnormally get executed on the other short-lived leader, which will certainly break the transaction model.

### PR DESCRIPTION
**Pull request by disksing**
Fix #1317 

<del>Still working on detailed test.</del>

### CODE REVIEW

File: src/server/kv.rs [hunk 1]
```diff
@@ -298,7 +298,10 @@ impl StoreHandler {
         resp.set_cmd_gc_resp(gc);
     }
 
-    pub fn on_request(&self, req: Request, on_resp: OnResponse) -> Result<()> {
+    pub fn on_request(&self, mut req: Request, on_resp: OnResponse) -> Result<()> {
+        // Go client may set it to 0 if using gogoprotobuf.
```
Comment (ngaut) [inline src/server/kv.rs]: Why go client needs term?

Comment (disksing) [inline src/server/kv.rs]: Context is sent from client.

**Comment by disksing** [comment]
Ready to review.
/cc @BusyJay @siddontang @ngaut @zhangjinpeng1987

File: src/server/kv.rs [hunk 2]
```diff
@@ -298,7 +298,10 @@ impl StoreHandler {
...
+        req.mut_context().clear_term();
```
Comment (siddontang) [inline src/server/kv.rs]: why clear term here?

Comment (disksing) [inline src/server/kv.rs]: gogoprotobuf will always set it, then we cannot use `has_term()`. See https://github.com/pingcap/kvproto/issues/97

File: src/storage/engine/raftkv.rs [hunk 3]
```diff
@@ -95,7 +95,7 @@ pub struct RaftKv<S: RaftStoreRouter + 'static> {
 
 enum CmdRes {
     Resp(Vec<Response>),
-    Snap(RegionSnapshot),
+    Snap((RegionSnapshot, u64)),
```
Comment (siddontang) [inline src/storage/engine/raftkv.rs]: I prefer using a field structure instead of tuple here.

File: src/storage/engine/raftkv.rs [hunk 4]
```diff
@@ -237,17 +240,23 @@ impl<S: RaftStoreRouter> Engine for RaftKv<S> {
         ASYNC_REQUESTS_COUNTER_VEC.with_label_values(&["snapshot", "all"]).inc();
         let req_timer = ASYNC_REQUESTS_DURATIONS_VEC.with_label_values(&["snapshot"]).start_timer();
 
+        let mut ctx2 = ctx.clone();
```
Comment (siddontang) [inline src/storage/engine/raftkv.rs]: does clone hurt the performance?

**Comment by disksing** [comment]
@siddontang PTAL. I've add `struct CbContext` to carry data that engines need to pass back to upper level.

File: src/server/coprocessor/endpoint.rs [hunk 5]
```diff
@@ -147,7 +147,7 @@ impl BatchRunnable<Task> for Host {
             let id = self.last_req_id;
             let sched = self.sched.clone();
             if let Err(e) = self.engine.async_snapshot(reqs[0].req.get_context(),
-                                                       box move |res| {
+                                                       box move |(_cb_ctx, res)| {
```
Comment (BusyJay) [inline src/server/coprocessor/endpoint.rs]: Use `_` directly.

File: src/raftstore/store/store.rs [hunk 6]
```diff
@@ -1100,6 +1100,12 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             return Err(box_err!("mismatch peer id {} != {}", peer.peer_id(), peer_id));
         }
 
+        let header = msg.get_header();
+        // If header's term is 2 verions behind current term, leadership may have been changed away.
+        if header.has_term() && peer.term() > header.get_term() + 1 {
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Use `get_term() != 0 && ...` instead.

File: src/storage/engine/mod.rs [hunk 7]
```diff
@@ -33,7 +33,17 @@ pub const TEMP_DIR: &'static str = "";
 const SEEK_BOUND: usize = 30;
 const DEFAULT_TIMEOUT_SECS: u64 = 5;
 
-pub type Callback<T> = Box<FnBox(Result<T>) + Send>;
+pub type Callback<T> = Box<FnBox((CbContext, Result<T>)) + Send>;
+
+pub struct CbContext {
```
Comment (BusyJay) [inline src/storage/engine/mod.rs]: Why define another `CbContext`? It change the API a lot. We have many context already, add another one just makes things complicated and confusing. I think it's ugly and unnecessary.

Comment (disksing) [inline src/storage/engine/mod.rs]: First of all, old API does not support underlying engine to return more data other than `Result`, we need to change it no matter we reuse `Context` or not.
In addition, the content of the two Context are different, and subsequent changes will make them more different. I see no reason to reuse it.

Comment (BusyJay) [inline src/storage/engine/mod.rs]: I'm not saying reuse the `Context`. You can check the `Result` to get the actual term.

Comment (disksing) [inline src/storage/engine/mod.rs]: In fact, my next PR need it return extra data no matter the result is Ok or Err.

**Comment by disksing** [comment]
PTAL @siddontang @BusyJay 
I see no better way other than add the new context. We might rethink it later.

**Comment by siddontang** [comment]
LGTM
PTAL @zhangjinpeng1987

File: src/raftstore/store/store.rs [hunk 8]
```diff
@@ -1100,6 +1100,12 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             return Err(box_err!("mismatch peer id {} != {}", peer.peer_id(), peer_id));
         }
 
+        let header = msg.get_header();
+        // If header's term is 2 verions behind current term, leadership may have been changed away.
+        if header.get_term() > 0 && peer.term() > header.get_term() + 1 {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: Why not use `peer.term() != hader.get_term()` directly?

Comment (disksing) [inline src/raftstore/store/store.rs]: term may increase without leader changed, we have already checked it is leader, so this case is okay.

**Comment by zhangjinpeng1987** [comment]
LGTM

File: tests/storage/test_raft_storage.rs [hunk 9]
```diff
@@ -59,3 +66,115 @@ fn test_raft_storage() {
...
+    ctx.set_region_epoch(region.get_region_epoch().clone());
+    ctx.set_peer(peers[0].clone());
+    ctx.set_term(term);
+
+    // Not leader.
+    cluster.must_transfer_leader(region.get_id(), peers[1].clone());
+    assert!(engine.write(&ctx, vec![]).is_err());
+    // Term not match.
+    cluster.must_transfer_leader(region.get_id(), peers[0].clone());
+    assert!(engine.write(&ctx, vec![]).is_err());
```
Comment (BusyJay) [inline tests/storage/test_raft_storage.rs]: How do you know the error due to term not matched?

File: src/storage/engine/mod.rs [hunk 10]
```diff
@@ -33,7 +33,17 @@ pub const TEMP_DIR: &'static str = "";
 const SEEK_BOUND: usize = 30;
 const DEFAULT_TIMEOUT_SECS: u64 = 5;
 
-pub type Callback<T> = Box<FnBox(Result<T>) + Send>;
+pub type Callback<T> = Box<FnBox((CbContext, Result<T>)) + Send>;
```
Comment (disksing) [inline src/storage/engine/mod.rs]: `wait_op!()` will be broken if not use tuple.

**Comment by disksing** [comment]
PTAL @BusyJay

File: tests/storage/test_raft_storage.rs [hunk 11]
```diff
@@ -59,3 +66,120 @@ fn test_raft_storage() {
...
+        })
+        .unwrap();
+    // Sleep a while, the prewrite should be blocked at snapshot stage.
+    thread::sleep(Duration::from_millis(50));
+    // Transfer leader twice, then unblock snapshot.
+    cluster.must_transfer_leader(region.get_id(), peers[1].clone());
+    cluster.must_transfer_leader(region.get_id(), peers[0].clone());
+    block.store(false, Ordering::SeqCst);
+
+    rx.recv().unwrap();
```
Comment (BusyJay) [inline tests/storage/test_raft_storage.rs]: Use `recv_timeout` instead.

File: src/raftstore/errors.rs [hunk 12]
```diff
@@ -114,6 +114,9 @@ quick_error!{
             description("region is stale")
             display("StaleEpoch {}", msg)
         }
+        StaleCommand {
```
Comment (BusyJay) [inline src/raftstore/errors.rs]: What will client do when encounter `StaleCommand`? Will it backoff?

Comment (disksing) [inline src/raftstore/errors.rs]: Yes. Backoff then retry.

Comment (BusyJay) [inline src/raftstore/errors.rs]: I don't think backoff is necessary here, because the new leader should be elected, it can retry immediately.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**